### DISCUSSION TITLE
Add `datetime`, `date`, and `time` platforms

### DISCUSSION THREAD

**Discussion by raman325** [home-assistant/architecture#811] [Entity Models]
# Status
- `date`: Merged
- `datetime`: In review
- `time`: Merged

## Background

All input_* have corresponding platforms for integrations to use except `input_datetime` and `input_text`. By creating a `datetime` platform, integrations can start creating entities for different use cases (see #812 for `text` proposal). Examples include:
- Setting a time for an action to be performed, e.g. setting a time for my robovac to start every day (the robovac has a built in time scheduler)
- Adjusting the schedule of tasks on an application that integrates with HA

## Proposal

~~Create a `datetime` platform~~

Create `datetime`, `date`, and `time` platforms

### State

string representation of datetime, time, or date depending on ~~`mode`~~ platform (we will use `native_value` and convert the value to state internally)

### Extra State Attributes

- ~~`mode`: one of [date, time, datetime] defaulting to `datetime`~~
- ~~`has_date`: true if entity has a date~~
- ~~`has_time`: true if entity has a time~~
- `year`: year of date (if date is included)
- `month`: month of date (if date is included)
- `day`: day of date (if date is included)
- ~~`timestamp`: timestamp representing the time (if time is included)~~
- `hour`: hour of time (if time is included, 0-23)
- `minute`: minute of time  (if time is included, 0-59)
- `second`: second of time (if time is included and `has_seconds` is true, 0-59)

### Services
- `datetime.set_value` - Sets a new value for the datetime (supports same attributes as the input_datetime.set_datetime service for date/times)
- `time.set_value` - Sets a new value for the time (supports same attributes as the input_datetime.set_datetime service for times)
- `date.set_value` - Sets a new value for the date (supports same attributes as the input_datetime.set_datetime service for dates)

### Integration must implement

- `native_value` property to be converted internally in the platform to state ~~based on `mode`~~
- ~~`async_set_datetime` or `set_datetime` method which has a single datetime/date/time value as a parameter. Platform internally converts service data to either `datetime`, `time`, or `date` base on `mode`~~
- `async_set_value` or `set_value` method

### Integration can implement

- ~~`mode` property~~

### Implementation

`datetime`: https://github.com/home-assistant/core/pull/81943 and https://github.com/home-assistant/frontend/pull/14391
`date`: https://github.com/home-assistant/core/pull/81948 and https://github.com/home-assistant/frontend/pull/14389
`time`: https://github.com/home-assistant/core/pull/81949 and https://github.com/home-assistant/frontend/pull/14390

**Comment by thecode** [comment]
I think it would be better to split this into two discussions

**Reply by raman325** [reply]
done

**Comment by frenck** [comment]
I do like this one, and it does match `input_datetime`, however, I wonder if it should be split into multiple entity types.

The problem with `input_datetime` at this point (IMHO), is the fact it supports multiple things:

- date
- time
- datetime

Therefore how and what you can set on it, is also a bit "magical" and often confusing. This is most apparent in the `input_datetime.set` service call:

![image](https://user-images.githubusercontent.com/195327/201029238-2460be9e-7113-4b40-b838-4b3f186a0214.png)

It just makes me wonder, if this should be split into multiple entity components?

**Reply by raman325** [reply]
I like this approach as the need to support multiple types in a single entity adds a lot of unnecessary complexity (see my reference implementation as an example).

**Reply by raman325** [reply]
made some updates to the proposal based on this idea

**Reply by wills106** [reply]
I'm looking forward to this being implemented.
This will be very useful for setting times on Solar Inverters with battery storage. At the moment I am having to use a mess of selects and numbers to be able to set the times for charger / discharge windows in my custom_component.

**Reply by raman325** [reply]
@frenck any additional thoughts about how to handle timezone with the `time` platform? At a minimum I think it would have to be a feature because there are use cases for timezoneless times. I'm struggling to find a use case for times with timezones since timezones affect the date in addition to the time. I am thinking we start with timezoneless and then we can add timezones in as an optional feature if/when there is a usecase that calls for it?

**Comment by frenck** [comment]
This architectural proposal has been fully implemented.

../Frenck
